File: amr/planning/dwa.py
"""Dynamic Window Approach local planner.

Samples a grid of (v, omega) commands within the velocity/acceleration-limited
dynamic window, rolls each out with a unicycle model, rejects rollouts that hit a
lethal costmap cell, and scores the survivors against a carrot point taken a fixed
arc length ahead on the global path.
"""
from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Tuple

import numpy as np

from amr.core.config import DwaConfig, RobotConfig
from amr.core.geometry import wrap_angle
from amr.core.types import Pose2D, Twist2D
from amr.planning.costmap import Costmap


@dataclass
class DwaResult:
    cmd: Twist2D
    trajectory: np.ndarray    # (K, 3) poses of the chosen rollout
    blocked: bool             # True if every sampled rollout collides


def carrot_point(path: np.ndarray, pose: Pose2D, lookahead: float) -> np.ndarray:
    """Point ``lookahead`` metres of arc length beyond the projection of the
    pose onto the path polyline; clamps to the final point (the goal)."""
    path = np.asarray(path, dtype=float)
    if len(path) == 1:
        return path[0].copy()
    p = np.array([pose.x, pose.y])

    # Find the closest point on any segment; record its cumulative arc length.
    best_d2 = float("inf")
    best_arc = 0.0
    arc = 0.0
    for i in range(len(path) - 1):
        a = path[i]
        b = path[i + 1]
        seg = b - a
        seg_len = math.hypot(seg[0], seg[1])
        if seg_len < 1e-12:
            continue
        t = float(np.dot(p - a, seg) / (seg_len * seg_len))
        t = min(1.0, max(0.0, t))
        proj = a + t * seg
        d2 = float((p[0] - proj[0]) ** 2 + (p[1] - proj[1]) ** 2)
        if d2 < best_d2:
            best_d2 = d2
            best_arc = arc + t * seg_len
        arc += seg_len

    target_arc = best_arc + lookahead

    # Walk forward along the polyline to the target arc length.
    arc = 0.0
    for i in range(len(path) - 1):
        a = path[i]
        b = path[i + 1]
        seg = b - a
        seg_len = math.hypot(seg[0], seg[1])
        if seg_len < 1e-12:
            continue
        if arc + seg_len >= target_arc:
            t = (target_arc - arc) / seg_len
            t = min(1.0, max(0.0, t))
            return a + t * seg
        arc += seg_len
    # Past the end of the path -> clamp to the final point.
    return path[-1].copy()


class DwaPlanner:
    def __init__(self, cfg: DwaConfig, robot: RobotConfig):
        self.cfg = cfg
        self.robot = robot

    def _rollout(self, pose: Pose2D, v: float, omega: float
                 ) -> np.ndarray:
        cfg = self.cfg
        n = int(round(cfg.sim_time / cfg.sim_dt))
        x, y, th = pose.x, pose.y, pose.theta
        out = np.empty((n, 3), dtype=float)
        dt = cfg.sim_dt
        for k in range(n):
            x += v * math.cos(th) * dt
            y += v * math.sin(th) * dt
            th = wrap_angle(th + omega * dt)
            out[k, 0] = x
            out[k, 1] = y
            out[k, 2] = th
        return out

    def compute(self, pose: Pose2D, vel: Twist2D, path: np.ndarray,
                costmap: Costmap) -> DwaResult:
        cfg = self.cfg
        robot = self.robot
        T = cfg.sim_time

        v_max = robot.max_lin_vel
        w_max = robot.max_ang_vel

        # Dynamic window (sim_time used as the reachability horizon).
        v_lo = max(0.0, vel.v - robot.max_lin_acc * T)
        v_hi = min(v_max, vel.v + robot.max_lin_acc * T)
        w_lo = max(-w_max, vel.omega - robot.max_ang_acc * T)
        w_hi = min(w_max, vel.omega + robot.max_ang_acc * T)

        carrot = carrot_point(path, pose, cfg.lookahead)
        goal = np.asarray(path, dtype=float)[-1]
        dist_to_goal = math.hypot(goal[0] - pose.x, goal[1] - pose.y)

        # Near-goal slowdown: cap sampled v.
        if dist_to_goal < cfg.lookahead:
            v_hi = min(v_hi, max(0.1, 0.7 * dist_to_goal))
            v_hi = max(v_hi, v_lo)

        v_grid = np.linspace(v_lo, v_hi, cfg.v_samples)
        w_grid = np.linspace(w_lo, w_hi, cfg.w_samples)

        # Collect surviving rollouts and their raw score components.
        trajs = []
        vels = []
        progress = []
        heading = []
        clearance = []
        velocity = []

        for v in v_grid:
            for w in w_grid:
                traj = self._rollout(pose, float(v), float(w))
                # Per-pose cost; reject if any pose is lethal.
                costs = np.empty(traj.shape[0], dtype=float)
                lethal = False
                for k in range(traj.shape[0]):
                    cck = costmap.cost_at_world(traj[k, 0], traj[k, 1])
                    costs[k] = cck
                    if cck >= 0.99:
                        lethal = True
                        break
                if lethal:
                    continue

                end = traj[-1]
                d_end = math.hypot(end[0] - carrot[0], end[1] - carrot[1])
                bearing = math.atan2(carrot[1] - end[1], carrot[0] - end[0])
                head_err = abs(wrap_angle(bearing - end[2]))
                clear = float(np.min(1.0 - costs))

                trajs.append(traj)
                vels.append((float(v), float(w)))
                progress.append(-d_end)
                heading.append(-head_err)
                clearance.append(clear)
                velocity.append(float(v) / v_max if v_max > 0 else 0.0)

        if not trajs:
            return DwaResult(Twist2D(0.0, 0.0),
                             np.empty((0, 3), dtype=float), blocked=True)

        progress = np.asarray(progress)
        heading = np.asarray(heading)
        clearance = np.asarray(clearance)
        velocity = np.asarray(velocity)

        def _norm(a: np.ndarray) -> np.ndarray:
            lo = float(a.min())
            hi = float(a.max())
            if hi - lo < 1e-12:
                return np.ones_like(a)
            return (a - lo) / (hi - lo)

        total = (cfg.w_progress * _norm(progress)
                 + cfg.w_heading * _norm(heading)
                 + cfg.w_clearance * _norm(clearance)
                 + cfg.w_velocity * _norm(velocity))

        best = int(np.argmax(total))
        bv, bw = vels[best]
        return DwaResult(Twist2D(bv, bw), trajs[best], blocked=False)

File: src/amr_planning/src/dwa.cpp
// SPDX-License-Identifier: Apache-2.0
#include "amr_planning/dwa.hpp"

#include <algorithm>
#include <cmath>
#include <limits>

#include "amr_core/geometry.hpp"

namespace amr_planning {

std::array<double, 2> carrot_point(const Path& path,
                                   const amr_core::Pose2D& pose,
                                   double lookahead) {
  if (path.size() == 1) {
    return path[0];
  }
  const double px = pose.x;
  const double py = pose.y;

  // Find the closest point on any segment; record its cumulative arc length.
  double best_d2 = std::numeric_limits<double>::infinity();
  double best_arc = 0.0;
  double arc = 0.0;
  for (std::size_t i = 0; i + 1 < path.size(); ++i) {
    const double ax = path[i][0];
    const double ay = path[i][1];
    const double bx = path[i + 1][0];
    const double by = path[i + 1][1];
    const double segx = bx - ax;
    const double segy = by - ay;
    const double seg_len = std::hypot(segx, segy);
    if (seg_len < 1e-12) {
      continue;
    }
    double t = ((px - ax) * segx + (py - ay) * segy) / (seg_len * seg_len);
    t = std::min(1.0, std::max(0.0, t));
    const double projx = ax + t * segx;
    const double projy = ay + t * segy;
    const double d2 = (px - projx) * (px - projx) + (py - projy) * (py - projy);
    if (d2 < best_d2) {
      best_d2 = d2;
      best_arc = arc + t * seg_len;
    }
    arc += seg_len;
  }

  const double target_arc = best_arc + lookahead;

  // Walk forward along the polyline to the target arc length.
  arc = 0.0;
  for (std::size_t i = 0; i + 1 < path.size(); ++i) {
    const double ax = path[i][0];
    const double ay = path[i][1];
    const double bx = path[i + 1][0];
    const double by = path[i + 1][1];
    const double segx = bx - ax;
    const double segy = by - ay;
    const double seg_len = std::hypot(segx, segy);
    if (seg_len < 1e-12) {
      continue;
    }
    if (arc + seg_len >= target_arc) {
      double t = (target_arc - arc) / seg_len;
      t = std::min(1.0, std::max(0.0, t));
      return {ax + t * segx, ay + t * segy};
    }
    arc += seg_len;
  }
  // Past the end of the path -> clamp to the final point.
  return path.back();
}

std::vector<std::array<double, 3>> DwaPlanner::rollout(
    const amr_core::Pose2D& pose, double v, double omega) const {
  const int n = static_cast<int>(std::lround(cfg_.sim_time / cfg_.sim_dt));
  double x = pose.x;
  double y = pose.y;
  double th = pose.theta;
  const double dt = cfg_.sim_dt;
  std::vector<std::array<double, 3>> out;
  out.reserve(static_cast<std::size_t>(std::max(0, n)));
  for (int k = 0; k < n; ++k) {
    x += v * std::cos(th) * dt;
    y += v * std::sin(th) * dt;
    th = amr_core::wrap_angle(th + omega * dt);
    out.push_back({x, y, th});
  }
  return out;
}

DwaResult DwaPlanner::compute(const amr_core::Pose2D& pose,
                              const amr_core::Twist2D& vel, const Path& path,
                              const Costmap& costmap) const {
  const double T = cfg_.sim_time;
  const double v_max = robot_.max_lin_vel;
  const double w_max = robot_.max_ang_vel;

  // Dynamic window (sim_time used as the reachability horizon).
  double v_lo = std::max(0.0, vel.v - robot_.max_lin_acc * T);
  double v_hi = std::min(v_max, vel.v + robot_.max_lin_acc * T);
  double w_lo = std::max(-w_max, vel.omega - robot_.max_ang_acc * T);
  double w_hi = std::min(w_max, vel.omega + robot_.max_ang_acc * T);

  const std::array<double, 2> carrot = carrot_point(path, pose, cfg_.lookahead);
  const std::array<double, 2> goal = path.back();
  const double dist_to_goal =
      std::hypot(goal[0] - pose.x, goal[1] - pose.y);

  // Near-goal slowdown: cap sampled v.
  if (dist_to_goal < cfg_.lookahead) {
    v_hi = std::min(v_hi, std::max(0.1, 0.7 * dist_to_goal));
    v_hi = std::max(v_hi, v_lo);
  }

  // np.linspace semantics: endpoints inclusive; a single sample yields v_lo.
  auto linspace = [](double lo, double hi, int num) {
    std::vector<double> g(static_cast<std::size_t>(std::max(0, num)));
    if (num == 1) {
      g[0] = lo;
      return g;
    }
    for (int i = 0; i < num; ++i) {
      g[static_cast<std::size_t>(i)] =
          lo + (hi - lo) * static_cast<double>(i) / static_cast<double>(num - 1);
    }
    return g;
  };

  const std::vector<double> v_grid = linspace(v_lo, v_hi, cfg_.v_samples);
  const std::vector<double> w_grid = linspace(w_lo, w_hi, cfg_.w_samples);

  // Collect surviving rollouts and their raw score components.
  std::vector<std::vector<std::array<double, 3>>> trajs;
  std::vector<std::array<double, 2>> vels;  // (v, w)
  std::vector<double> progress, heading, clearance, velocity;

  for (double v : v_grid) {
    for (double w : w_grid) {
      auto traj = rollout(pose, v, w);
      // Per-pose cost; reject if any pose is lethal.
      bool lethal = false;
      double min_one_minus_cost = std::numeric_limits<double>::infinity();
      for (const auto& p : traj) {
        const double cck = costmap.cost_at_world(p[0], p[1]);
        if (cck >= 0.99) {
          lethal = true;
          break;
        }
        min_one_minus_cost = std::min(min_one_minus_cost, 1.0 - cck);
      }
      if (lethal) {
        continue;
      }

      const auto& end = traj.back();
      const double d_end = std::hypot(end[0] - carrot[0], end[1] - carrot[1]);
      const double bearing =
          std::atan2(carrot[1] - end[1], carrot[0] - end[0]);
      const double head_err = std::abs(amr_core::wrap_angle(bearing - end[2]));
      const double clear = min_one_minus_cost;

      trajs.push_back(std::move(traj));
      vels.push_back({v, w});
      progress.push_back(-d_end);
      heading.push_back(-head_err);
      clearance.push_back(clear);
      velocity.push_back(v_max > 0.0 ? v / v_max : 0.0);
    }
  }

  if (trajs.empty()) {
    DwaResult r;
    r.cmd = amr_core::Twist2D{0.0, 0.0};
    r.blocked = true;
    return r;
  }

  // Min-max normalisation: constant component -> all ones (matches Python).
  auto norm = [](const std::vector<double>& a) {
    double lo = std::numeric_limits<double>::infinity();
    double hi = -std::numeric_limits<double>::infinity();
    for (double x : a) {
      lo = std::min(lo, x);
      hi = std::max(hi, x);
    }
    std::vector<double> out(a.size());
    if (hi - lo < 1e-12) {
      std::fill(out.begin(), out.end(), 1.0);
      return out;
    }
    for (std::size_t i = 0; i < a.size(); ++i) {
      out[i] = (a[i] - lo) / (hi - lo);
    }
    return out;
  };

  const std::vector<double> np_progress = norm(progress);
  const std::vector<double> np_heading = norm(heading);
  const std::vector<double> np_clearance = norm(clearance);
  const std::vector<double> np_velocity = norm(velocity);

  int best = 0;
  double best_total = -std::numeric_limits<double>::infinity();
  for (std::size_t i = 0; i < trajs.size(); ++i) {
    const double total = cfg_.w_progress * np_progress[i] +
                         cfg_.w_heading * np_heading[i] +
                         cfg_.w_clearance * np_clearance[i] +
                         cfg_.w_velocity * np_velocity[i];
    if (total > best_total) {
      best_total = total;
      best = static_cast<int>(i);
    }
  }

  DwaResult r;
  r.cmd = amr_core::Twist2D{vels[static_cast<std::size_t>(best)][0],
                            vels[static_cast<std::size_t>(best)][1]};
  r.trajectory = trajs[static_cast<std::size_t>(best)];
  r.blocked = false;
  return r;
}

}  // namespace amr_planning
